File: src/StaticDoublyLinkedList.zig
const std = @import("std");
const expect = std.testing.expect;
const expectEqual = std.testing.expectEqual;
const expectError = std.testing.expectError;

pub fn StaticDoublyLinkedList(comptime T: type, capacity: usize) type {
    const BitSet = std.StaticBitSet(capacity);

    return struct {
        const Self = @This();

        const ListNode = struct {
            previous: ?*ListNode = null,
            next: ?*ListNode = null,
            value: T,
        };

        values: [capacity]ListNode = undefined,
        capacity: usize = capacity,
        first: ?*ListNode = null,
        last: ?*ListNode = null,
        free: BitSet = BitSet.initFull(),
        size: usize = 0,

        pub fn append(list: *Self, value: T) !*ListNode {
            if (list.size == capacity) return error.OutOfMemory;

            const index = list.free.findFirstSet().?;
            list.free.unset(index);

            const new_node = &list.values[index];
            new_node.value = value;
            new_node.previous = list.last;
            new_node.next = null;

            if (list.last) |last| {
                last.next = new_node;
            } else {
                list.first = new_node;
            }
            list.last = new_node;

            list.size += 1;
            return new_node;
        }

        pub fn prepend(list: *Self, value: T) !*ListNode {
            if (list.size == capacity) return error.OutOfMemory;

            const index = list.free.findFirstSet().?;
            list.free.unset(index);

            const new_node = &list.values[index];
            new_node.value = value;
            new_node.previous = null;
            new_node.next = list.first;

            if (list.first) |first| {
                first.previous = new_node;
            } else {
                list.last = new_node;
            }
            list.first = new_node;

            list.size += 1;
            return new_node;
        }

        pub fn insertAfter(list: *Self, node: *ListNode, value: T) !*ListNode {
            if (list.size == capacity) return error.OutOfMemory;

            const index = list.free.findFirstSet().?;
            list.free.unset(index);

            const new_node = &list.values[index];
            new_node.value = value;
            new_node.previous = node;

            if (node.next) |next_node| {
                new_node.next = next_node;
                next_node.previous = new_node;
            } else {
                new_node.next = null;
                list.last = new_node;
            }
            node.next = new_node;

            list.size += 1;
            return new_node;
        }

        pub fn remove(list: *Self, node: *ListNode) void {
            if (node.previous) |prev_node| {
                prev_node.next = node.next;
            } else {
                list.first = node.next;
            }

            if (node.next) |next_node| {
                next_node.previous = node.previous;
            } else {
                list.last = node.previous;
            }

            const index = (@intFromPtr(node) - @intFromPtr(&list.values[0])) / @sizeOf(ListNode);
            list.free.set(index);
            list.size -= 1;
        }

        pub fn pop(list: *Self) ?T {
            if (list.last == null) return null;

            const last_node = list.last.?;
            const value = last_node.value;

            list.remove(last_node);
            return value;
        }

        pub fn popFirst(list: *Self) ?T {
            if (list.first == null) return null;

            const first_node = list.first.?;
            const value = first_node.value;

            list.remove(first_node);
            return value;
        }

        pub fn length(list: *Self) usize {
            return capacity - list.free.count();
        }
    };
}

test "StaticDoublyLinkedList" {
    const List = StaticDoublyLinkedList(u32, 4);
    var list = List{};
    const n1 = try list.append(10);
    const n2 = try list.append(20);
    const n3 = try list.prepend(5);
    const n4 = try list.append(30);

    try expect(list.size == 4);
    try expectError(error.OutOfMemory, list.append(40));

    try expectEqual(list.first, n3);
    try expectEqual(list.last, n4);
    try expectEqual(null, n4.next);
    try expectEqual(n2, n1.next.?);
    try expectEqual(n4, n2.next.?);
    try expectEqual(n1, n3.next.?);

    list.remove(n3);
    try expectEqual(n4, n2.next.?);
    try expectEqual(n2, n4.previous.?);

    try expectEqual(3, list.size);
}

File: ts/src/staticDoublyLinkedList.ts
import { nullthrows as nt } from "./nullthrows";

export class StaticDoublyLinkedList {
  private readonly values: Int32Array;
  private readonly next: Int32Array;
  private readonly prev: Int32Array;
  private readonly inUse: Uint8Array;
  private readonly freeStack: Int32Array;
  private freeTop: number;

  private firstIndex = -1;
  private lastIndex = -1;
  size = 0;

  constructor(private readonly capacity: number) {
    if (!Number.isInteger(capacity) || capacity <= 0) {
      throw new RangeError(
        "StaticDoublyLinkedList: capacity must be positive integer",
      );
    }
    this.values = new Int32Array(capacity);
    this.next = new Int32Array(capacity).fill(-1);
    this.prev = new Int32Array(capacity).fill(-1);
    this.inUse = new Uint8Array(capacity);
    this.freeStack = new Int32Array(capacity);
    this.freeTop = capacity;
    for (let i = 0; i < capacity; i += 1) {
      this.freeStack[i] = capacity - 1 - i;
    }
  }

  private allocNode(): number {
    if (this.freeTop === 0) {
      throw new RangeError("StaticDoublyLinkedList overflow");
    }
    const index = nt(this.freeStack[--this.freeTop]);
    this.inUse[index] = 1;
    this.next[index] = -1;
    this.prev[index] = -1;
    return index;
  }

  private freeNode(index: number): void {
    this.inUse[index] = 0;
    this.next[index] = -1;
    this.prev[index] = -1;
    this.freeStack[this.freeTop++] = index;
  }

  get first(): number {
    return this.firstIndex;
  }

  get last(): number {
    return this.lastIndex;
  }

  append(value: number): number {
    const index = this.allocNode();
    this.values[index] = value;

    if (this.lastIndex !== -1) {
      this.next[this.lastIndex] = index;
      this.prev[index] = this.lastIndex;
    } else {
      this.firstIndex = index;
    }

    this.lastIndex = index;
    this.size += 1;
    return index;
  }

  prepend(value: number): number {
    const index = this.allocNode();
    this.values[index] = value;

    if (this.firstIndex !== -1) {
      this.prev[this.firstIndex] = index;
      this.next[index] = this.firstIndex;
    } else {
      this.lastIndex = index;
    }

    this.firstIndex = index;
    this.size += 1;
    return index;
  }

  insertAfter(nodeIndex: number, value: number): number {
    this.assertInUse(nodeIndex);

    const index = this.allocNode();
    this.values[index] = value;

    const nextIndex = nt(this.next[nodeIndex]);
    this.next[nodeIndex] = index;
    this.prev[index] = nodeIndex;

    if (nextIndex !== -1) {
      this.prev[nextIndex] = index!;
      this.next[index] = nextIndex;
    } else {
      this.lastIndex = index;
    }

    this.size += 1;
    return index;
  }

  remove(nodeIndex: number): void {
    this.assertInUse(nodeIndex);

    const prevIndex = nt(this.prev[nodeIndex]);
    const nextIndex = nt(this.next[nodeIndex]);

    if (prevIndex !== -1) {
      this.next[prevIndex] = nextIndex!;
    } else {
      this.firstIndex = nextIndex;
    }

    if (nextIndex !== -1) {
      this.prev[nextIndex] = prevIndex!;
    } else {
      this.lastIndex = prevIndex;
    }

    this.freeNode(nodeIndex);
    this.size -= 1;
  }

  pop(): number | null {
    if (this.lastIndex === -1) return null;
    const value = this.values[this.lastIndex];
    this.remove(this.lastIndex);
    return value ?? null;
  }

  popFirst(): number | null {
    if (this.firstIndex === -1) return null;
    const value = this.values[this.firstIndex];
    this.remove(this.firstIndex);
    return value ?? null;
  }

  length(): number {
    return this.size;
  }

  valueOf(index: number): number {
    this.assertInUse(index);
    return this.values[index]!;
  }

  nextOf(index: number): number {
    this.assertInUse(index);
    return this.next[index]!;
  }

  prevOf(index: number): number {
    this.assertInUse(index);
    return this.prev[index]!;
  }

  private assertInUse(index: number): void {
    if (index < 0 || index >= this.capacity || this.inUse[index] === 0) {
      throw new RangeError("StaticDoublyLinkedList: invalid node");
    }
  }
}
